### DISCUSSION TITLE
Signature generation

### DISCUSSION THREAD

**Discussion by boan-jfm** [libvips/php-vips#222] [Q&A] [answered]
Hi

Does php-vips support [signature generation](https://www.php.net/manual/en/imagick.getimagesignature.php) like imagemagick?

> SignatureImage() computes a message digest from an image pixel stream with an implementation of the NIST SHA-256 Message Digest algorithm. This signature uniquely identifies the image and is convenient for determining if an image has been modified or whether two images are identical.
> -- https://imagemagick.org/api/signature.php

If not, are there alternatives to generate a likewise signature based on pixeldata only?

**Comment by jcupitt** [comment, marked as answer]
Hi again, no there's nothing like that. I think I'd just shell out to `md5sum` on the source file.

**Reply by boan-jfm** [reply]
The issue with doing that is that the hash will be based on the metadata in the image file as well. We use it to check if an image has been imported before.

Thank you for your time 👍

**Reply by jcupitt** [reply]
If you just need to detect duplicates, I'd use something like dhash. In ruby it's:

```ruby
def hash_thumbnail filename
    image = Vips::Image.thumbnail filename, 9, height: 8, size: :force
    image = image.colourspace 'b-w'
    image = image.cast('int').conv([1, -1], precision: :integer)
    image = image.crop(1, 0, 8, 8) > 0
    array = (image / 255).cast('uchar').to_a
    array.join.to_i(2)
end
```

So that's shrinking to 8x8 mono pixels, finding the average, then setting bits to 0 or 1 depending on whether they are brighter or darker than that. Now you have 64 bits, or a long int. It's a 'structural' hash, so similar images will hash to similar values. It only takes a few ms to compute, and because it's just an int, you can search a large image DB for a hit very quickly.

Store that 64-bit int in your DB and look for matches on import. If you get a hit, do a more careful test (perhaps thumbnail, then look for sum of squares of differences less than 100?) just on the matching pair.

**Reply by jcupitt** [reply]
Oh heh that's a slightly different one. Anyway, some kind of structure hash like that.

**Reply by jcupitt** [reply]
... you can also compute the hamming distance between pairs of hashes and use that to rank searches for similar images, which can be useful.

**Reply by boan-jfm** [reply]
I must admit, I am not familiar with Ruby, and I do not understand how you can divide an object by 255 (`image / 255`). PHP-VIPS seems to have a different signature for the conv method; it uses an image object rather than an array:

> Image conv(Image $mask, array $options = []) Convolution operation.
> -- https://github.com/libvips/php-vips/blob/v2.3.0/src/ImageAutodoc.php#L153

(libvips: https://www.libvips.org/API/current/libvips-convolution.html#vips-conv)

I have tried to convert it to PHP:

```php
        $image = Image::thumbnail($test_image_path, 9, ['height' => 8, 'size' => 'force']);
        $image = $image->colourspace('b-w');
        $image = $image->cast('int')->conv($image, ['precision' => 'integer']);
        $image = $image->crop(0, 0, 8, 8);
        $array = $image->cast('uchar')->writeToArray();
        $binaryString = implode('', $array);
        $hash = bindec($binaryString);
```

Without the 'conv' call, it executes with a huge number as a result

**Reply by jcupitt** [reply]
In php it'd be:

```php
#!/usr/bin/env php
<?php

require __DIR__ . '/vendor/autoload.php';
use Jcupitt\Vips;

function dhash($filename)
{
    $image = Vips\Image::thumbnail($filename, 9, [
        'height' => 8,
        'size' => 'force'
    ]);
    $image = $image->colourspace('b-w');
    $image = $image->cast('int')->conv([[-1,  1]], ['precision' => 'integer']);
    $image = $image->crop(1, 0, 8, 8)->more(0)->divide(255)->cast('uchar');
    $array = $image->writeToArray();
    $binaryString = implode('', $array);
    $hash = bindec($binaryString);

    return $hash;
}

for ($i = 1; $i < count($argv); $i++)
{
    $hash = dhash($argv[$i]);
    echo "${argv[$i]}: $hash\n";
}
```

Though it looks like php doesn't have 64-bit ints, unfortunately.

**Reply by jcupitt** [reply]
> I do not understand how you can divide an object by 255 (image / 255)

That returns a new image with every pixel divided by 255. php doesn't have operator overloads, so you have to write `->divide(255)` instead.

> PHP-VIPS seems to have a different signature for the conv method; it uses an image object rather than an array:

You can pass an array to `conv` in PHP as well. It'll get cast into an image for you.

**Reply by boan-jfm** [reply]
Thank you, it seems to  work.

Just curious.

1) In Ruby, you provide the matrix [1, -1], whereas in PHP, you provide [[-1, 1]] as an argument. In PHP, it's a nested array, and the numbers appear to be switched?

2) Why is it cropped from the left by 1 and not 0, with the top being 0 and not 1 when calling the crop method?


> The size of an int is platform-dependent, although a maximum value of about two billion is the usual value (that's 32 bits signed). 64-bit platforms usually have a maximum value of about 9E18. PHP does not support unsigned ints. int size can be determined using the constant PHP_INT_SIZE, maximum value using the constant PHP_INT_MAX, and minimum value using the constant PHP_INT_MIN.
> https://www.php.net/manual/en/language.types.integer.php

**Reply by jcupitt** [reply]
1. It's computing taking the left-right pixel difference, so it doesn't really matter which is 1 and which is -1. It's supposed to cast a 1D array into a 2D array for you, but php-vips doesn't seem to do this (argh!), so I made a 2D array.
2. It's computing left-right difference, so it shrinks to 9 pixels across, then drops a column. Again, it doesn't really matter which column you drop. Probably dropping 0 is best.
